src/Uno.Foundation/Runtime.wasm.cs

```diff
@@ -301,6 +305,78 @@ public static string InvokeJSWithInterop(FormattableString formattable)
 			return InvokeJS(command);
 		}
 
+		private static readonly Dictionary<long, TaskCompletionSource<string>> _asyncWaitingList = new Dictionary<long, TaskCompletionSource<string>>();
+
+		private static long _nextAsync;
+
+		/// <summary>
+		/// Invoke async javascript code.
+		/// </summary>
+		/// <remarks>
+		/// The javascript code is expected to return a Promise&lt;string&gt;
+		/// </remarks>
+		public static Task<string> InvokeAsync(string promiseCode)
+		{
+			var id = Interlocked.Increment(ref _nextAsync);
+
+			var tcs = new TaskCompletionSource<string>();
+
+			lock (_asyncWaitingList)
+			{
+				_asyncWaitingList[id] = tcs;
+			}
+
+			var js = new[]
+			{
+				"const __f = ()=>",
+				promiseCode,
+				";\nUno.UI.Interop.AsyncInteropHelper.Invoke(",
+				id.ToStringInvariant(),
+				", __f);"
+			};
+
+			try
+			{
+
+				WebAssemblyRuntime.InvokeJS(string.Concat(js));
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<string>(ex);
```

Comment:
Seems like `tcs` gets leaked in this case? Shouldn't it be removed from the list?

Reply:
Oooh you're right.